value/src/main/java/com/google/auto/value/AutoValueExtension.java

```diff
@@ -0,0 +1,100 @@
+package com.google.auto.value;
+
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * An AutoValueExtension allows for extra functionality to be created during the generation
+ * of an AutoValue class.
+ *
+ * Extensions are discovered at compile time using the {@link java.util.ServiceLoader} APIs,
+ * allowing them to run without any additional annotations.
+ *
+ * Extensions can extend the AutoValue implementation by generating subclasses of the AutoValue
+ * generated class. It's not guaranteed that an Extension's generated class will be the final
+ * class in the inheritance hierarchy, unless it's {@link #mustBeAtEnd} method returns true,
+ * and only one Extension at a time can return true for a given context.  Only generated classes
+ * that will be the final class in the inheritance hierarchy can be final.
+ *
+ * Each Extension must also be sure to generate a constructor with arguments corresponding to
+ * all properties in {@link com.google.auto.value.AutoValueExtension.Context#properties()}, in
+ * order.  This constructor must have at least package visibility.
+ */
+public interface AutoValueExtension {
+
+  /**
+   * The context of the generation cycle.
+   */
+  public interface Context {
+
+    /**
+     * The processing environment of this generation cycle.
+     *
+     * @return The {@link javax.annotation.processing.ProcessingEnvironment} of this generation cycle.
```

Comment:
I think just `{@link ProcessingEnvironment}` would be fine, or just omit `{@link}` since it is the return type of the method.